From 2306b6edbb55a8a2a68e95ad55935ae1a5386b6e Mon Sep 17 00:00:00 2001
From: Daniel Murphy <dmurph@chromium.org>
Date: Tue, 11 Jun 2024 16:10:00 +0000
Subject: [PATCH] [PWA] Fix manifest badmessage due to overriding to add icons.

The override infrastructure was false triggering a badmessage when the
site had a malformed manifest, or other problems happened like the
frame being destroyed.

This is a minimal patch to prevent this badmessage from occurring.

(cherry picked from commit 89edb4438eac0b665578103090256fe2f8d1c686)

Fixed: 332810010
Change-Id: Ia34605e66489858b78c8ef551c617eeb53d73e10
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/5598018
Commit-Queue: Daniel Murphy <dmurph@chromium.org>
Reviewed-by: Marijn Kruisselbrink <mek@chromium.org>
Cr-Original-Commit-Position: refs/heads/main@{#1310232}
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/5618158
Reviewed-by: Dibyajyoti Pal <dibyapal@chromium.org>
Cr-Commit-Position: refs/branch-heads/6478@{#1439}
Cr-Branched-From: e6143acc03189c5e52959545b110d6d17ecd5286-refs/heads/main@{#1300313}
---

diff --git a/content/browser/manifest/manifest_manager_host.cc b/content/browser/manifest/manifest_manager_host.cc
index bcb8f5f..24e3450 100644
--- a/content/browser/manifest/manifest_manager_host.cc
+++ b/content/browser/manifest/manifest_manager_host.cc
@@ -108,10 +108,6 @@
     const GURL& url,
     blink::mojom::ManifestPtr manifest) {
   DCHECK_CURRENTLY_ON(BrowserThread::UI);
-  GetContentClient()->browser()->MaybeOverrideManifest(
-      &page().GetMainDocument(), manifest);
-  auto callback = std::move(*callbacks_.Lookup(request_id));
-  callbacks_.Remove(request_id);
   // Note: An empty manifest signifies an error for callers. If a bad message
   // occurs, this can be used to not drop the callback & signify an error.
   if (!manifest) {
@@ -137,6 +133,17 @@
                         valid_to_string(scope_valid), ")."}));
     }
   }
+  // Empty manifests means it failed to parse or an unresolvable problem like
+  // the frame is destroying. Since AppBannerManager completely ignores these
+  // manifests, avoid overriding them as well to prevent the overriding
+  // infrastructure from seeing manifests without the default members.
+  if (!blink::IsEmptyManifest(manifest)) {
+    GetContentClient()->browser()->MaybeOverrideManifest(
+        &page().GetMainDocument(), manifest);
+  }
+  auto callback = std::move(*callbacks_.Lookup(request_id));
+  callbacks_.Remove(request_id);
+
   std::move(callback).Run(url, std::move(manifest));
 }
 
